### ISSUE TITLE
Make the compiler package aware

### ISSUE THREAD

**Issue by garyb** [purescript/purescript#3493]
Opening an issue as a place to discuss this. There was a lot of chat on Slack about things related to #2437 yesterday, and this idea was floated again. I think it was @natefaubion and @kRITZCREEK in particular that had some points of why this might be preferable, so please reiterate below :wink:

**Comment by kRITZCREEK** [comment]
Quoted from Slack:

To me making the compiler package aware is making sure we can introduce more features without making a breaking release after declaring 1.0.0.

For example:
1. Make a module (or explicit exported declarations) package visible, so we can avoid the .Internal nonsense for declarations that shouldn't be exposed and just want to be shared between different modules in a package
2. Think about raising the compilation unit level from module to package
  - to allow for more aggressive monomorphization for production builds
  - (other things I'm sure @rightfold has more opinions on)

**Comment by kRITZCREEK** [comment]
I don't think I'd want to go that far, especially in dev-builds the fine grained dependency tracking and good parallelism we get from our per-module compilation unit gives us acceptable compile times. So I don't think we should allow things that force us to raise the compilation unit to a package.

**Comment by hdgarrood** [comment]
One thing I’d really like to do is to make it so that you can’t import transitive dependencies directly without declaring that you’re using them; then we don’t need to consider dropping a dependency as a breaking change.

**Comment by chexxor** [comment]
If the compiler is package-aware, how aware of package distribution should it be? I think the compiler just needs to know the directories on your system in which you're keeping packages. Package location, publishing, and retrieval can be figured out later, then. 

Should package names be globally unique? Not when it comes to distribution and retrieval, but only when the compiler is using them.

**Comment by rightfold** [comment]
@chexxor The trivial mechanism would be to give it a mapping from package name to path on the command-line. Then, no directories have to be searched, so it stays very simple. This is at the cost of having to pass more arguments to the compiler.

Still, the compiler is rarely invoked manually. I think baking conventions into the compiler is a bad idea; it makes build systems harder to configure. The configuration of the compiler should be as explicit as possible.

**Comment by hdgarrood** [comment]
> The trivial mechanism would be to give it a mapping from package name to path on the command-line. Then, no directories have to be searched, so it stays very simple.
> ```
> purs compile --package prelude packages/purescript-prelude/output \
>             --package maybe packages/purescript-maybe/output \
>             src/**/*.purs
> ```

Unfortunately I don't think this is workable because of the length of the CLI invocations it would lead to. Windows refuses to execute commands which pass a certain length; I don't remember precisely what it is, but I'm pretty certain it's small enough that this would cause problems.

Also, I think it's a good thing that packages can't currently have their own build scripts. I'd very much prefer that to remain the case.

**Comment by chexxor** [comment]
@hdgarrood that problem can be resolved by allowing the compiler to accept a configuration file instead of command-line flags. I haven't looked for existing discussion for that idea, but it seems like an obvious benefit-only change to me.

**Comment by hdgarrood** [comment]
> it seems like an obvious benefit-only change

I very much disagree: the larger your API surface area is, the more effort is required to understand, document, or debug it. Of course we will have to expand the compiler's CLI to some degree if we do end up doing this, but if we do, I would want to ensure that we're not doing so any more than is absolutely necessary.

**Comment by hdgarrood** [comment]
Another thing we'd want to do if we were going to do this is to address #2477 at the same time (allow building against dependencies which have already been compiled). In an ideal world you'd only need to build a particular library with a particular compiler version once.

I suppose we might want to come up with a spec for a package database, which would be a bunch of packages laid out on disk in a particular way (including all source files and compilation results). Then when you're building your own project, you'd give `purs` a path to a package database, and a list of all the dependencies you're using with their versions. This information is very similar to the contents of resolutions file you currently have to pass to `purs publish`, so presumably we'd also want to have `purs publish` understand the same format too (see #3499 for additional context).

**Comment by chexxor** [comment]
This thread needs a definition of "package" at some point. Perhaps we need to brainstorm more problems that might best be solved with the notion of a "package". Based on this thread, current motivations: "scope of a compilation unit" and "dependency visibility to package consumers". For contrast, [Go-lang's "Packages" docs](https://www.golang-book.com/books/intro/11) say that the primary purpose of packages is to organize code and reduce name collisions.

Suggestions for other language-related topics a package might want to affect:

- **Encapsulation**: "In Go if something starts with a capital letter that means other packages (and programs) are able to see it." [Go-lang's "Packages" docs](https://www.golang-book.com/books/intro/11)

- **Symbol-to-File or Symbol-to-Directory**: ECMAScript "import maps" overload module definition mappings to support defining packages/directories, using a trailing slash: https://github.com/WICG/import-maps#packages-via-trailing-slashes. For contrast, the package import syntax I've seen suggested in other PS forums uses quotes or a colon to delimit the package from the module inside it: `import "purescript-strings".Data.String` or `import purescript-strings:Data.String`.

- **Location-encoded Names vs Symbolic Names**: Also in ECMAScript modules, the "location" of the module is specified out-of-band, in an "import-maps" configuration, while the "name" of the module is used in ECMAScript programs. This is analogous to what we're proposing for resolving the packages containing PureScript modules. For contrast, [Dhall-lang](https://dhall-lang.org/) uses fully-resolvable names, `let generate = https://prelude.dhall-lang.org/List/generate`.

- **Project Upgradability/Versioning**:  Relevant to "quality of life" in long-lived, big projects, there are practical benefits to supporting multiple versions of a module in a single project, as seen in the following article on thoughts affecting the Go-lang package re-design.

> I was also surprised to realize that there is a second, independent logical route to semantic import versioning: gradual code repair or partial code upgrades. In a large program, it’s unrealistic to expect all packages in the program to update from v1 to v2 of a particular dependency at the same time. Instead, it must be possible for some of the program to keep using v1 while other parts have upgraded to v2. But then the program’s build, and the program’s final binary, must include both v1 and v2 of the dependency. Giving them the same import path would lead to confusion, violating what we might call the import uniqueness rule: different packages must have different import paths. The only way to have partial code upgrades, import uniqueness, and semantic versioning is to adopt semantic import versioning as well.
> ~ https://blog.golang.org/versioning-proposal

- **Package Content Translations**: A personal thought: If I want my project's team to use `data Result a = Just a | Empty` but the core libraries use `data Maybe a = Just a | Nothing`, should a package mapping support renames like this? This is more practical when renaming, for example, Functor to Mappable and Monad to Bindable.

**Comment by chexxor** [comment]
If "package" implies lots of things, such as 1) Compilation boundary, 2) Namespace boundary, 3) Encapsulation boundary, 4) Dependency boundary, it would make "package" strictly less useful because the things which compose it are not able to be used independently of it. This is similar to a function having an `Integer` parameter be strictly "less useful" than one with a `Semiring a => a` parameter.

I mention this because a single project serving as an application rather than a library might find use of the *compilation boundary* but not want a *namespace boundary*.

**Comment by hdgarrood** [comment]
I'd suggest we take the simplest possible workable definition for now: as far as the compiler is concerned, a package is basically a record with the following fields:
- package name
- A collection of modules
- A list of packages it depends on, which are visible from inside the package

I'm not yet sure if the compiler should know about package versions. I'm leaning towards no, but I think this will be clearer after we have a proof-of-concept.

**Comment by hdgarrood** [comment]
@chexxor To respond to your points:

- encapsulation: if we are going to make it possible to have modules in a package which are not exposed by that package (I know this is something people have differing views on), I’d want to keep that as simple as possible, so I’d suggest that as far as a package is concerned, a module is either externally visible or it isn’t. Individual modules can of course decide for themselves which declarations they expose (this is already the case).
- on module resolution: Currently there is no mechanism by which a module name can be used to locate the corresponding source file on the disk. This might be a useful property to have but I think it would be a fairly seismic change so I’d suggest doing it separately from this (if at all).
- On project upgradability: I still think being able to depend on multiple versions of a package would be super hard to implement safely and I don’t think it’s clear that the extra flexibility it gives you

**Comment by hdgarrood** [comment]
Oops, premature submission there. Re depending on multiple versions: I don’t think it’s clear that the extra flexibility that you get from being able to depend on multiple versions makes up for the complications which come along with it, not to mention the amount of work it would take to implement it.
I’m not keen on package content translations. Note that you can use existing language features to rename things, like type synonyms for types or just `foo = bar` to define a value `foo` which is an alias for `bar`. Type synonyms don’t work on class names now but hopefully one day we will get around to doing constraint kinds, which would imply being able to use class names in type synonyms.

**Comment by denibertovic** [comment]
@hdgarrood What's currently stopping us from getting #2477 resolved? Is this currently being worked on or planned for the future? If not, what can I do to help move it along?

**Comment by hdgarrood** [comment]
#2477 should probably be subsumed into this issue; since I think these things are related - dependencies you might want to reuse compile artifacts for should almost always be in separate packages themselves, and if they aren't it's probably good practice to break them out into a separate package anyway.

For this issue, I think what we are missing is a specification and/or implementation plan. The best way to help right now is probably to come up with a slightly less vague suggestion of how this might work based on what people have said up-thread; I expect that would revitalize the discussion.

**Comment by hdgarrood** [comment]
I'll cross-post a comment I just posted on the spago issue https://github.com/spacchetti/spago/issues/527 here, as I think the justification for why I'm keen on this isn't sufficiently clear right now:

> Essentially, my view is that it makes sense to consider two types of inputs to builds: published packages at specific versions, which don't change and which are often depended on across different projects, and project-local code, which changes all the time and isn't depended on across different projects. These should be treated differently by the build system: because project-local files are liable to change often, we should be checking their hashes and/or timestamps on every build, but there's little sense in checking whether published packages have changed since the last build. In fact, right now, the compiler has to not only check timestamps of all source from your dependencies on every build, it also has to parse all of their imports and construct a module graph including all modules from all dependencies. If we had a concept of a package database which the compiler could use to store precompiled packages, we could skip this step, and the compiler would only have to read and parse files which were local to the project being built. Personally, I don't really see a way to a build system which I consider sufficiently fast without addressing this.

**Comment by hdgarrood** [comment]
@natefaubion Do you think it would make sense to say that there should be a reasonably clear specification for something before it can be moved into "intend to implement"? Putting this issue into that milestone in its current state (i.e. lacking a spec) feels a bit dodgy to me, personally.

**Comment by natefaubion** [comment]
Intend to implement doesn't mean ready to implement. The description of the milestone is that a maintainer has expressed a good-faith desire to see the feature through implementation and release. That is, it's going to happen sooner or later. I'm happy to remove it though if that was presumptuous.

**Comment by hdgarrood** [comment]
That makes sense. I am up for seeing the feature through implementation and release; I guess I just want to reserve the right to change my mind if we uncover a fatal issue in the process of assembling a spec.

**Comment by natefaubion** [comment]
Sure, that can change at any time. It's meant to provide visibility into high priority features to promote discussion and make users aware of what future releases of PureScript might contain.

**Comment by hdgarrood** [comment]
Ok cool - I'm good with leaving it as is then :)

**Comment by joneshf** [comment]
There's been a lot of talk of the benefits of packages, but have we considered that creating an idea of packages of modules is covering up a smell?

The reason we have packages in the ecosystem is because people want to distribute and consume multiple modules in a clump. But, why do we want to encourage people to distribute and consume multiple modules in a clump? Why don't we want people to distribute and consume single modules? Why don't we want to foster an ecosystem where there's one module to import to solve a specific problem? Why can't we push people to create solutions that don't require a `Types` module to exist? Why don't we want multiple PureScript files to work together as a single unit that you cannot "see through?"

I guess what I'm saying is that whether there are modules or packages (exclusively) shouldn't matter so much. But we ought to be really concerned that we want to be in a world with both at the same time that anybody can interact with.

N.B. The above questions are rhetorical. Please don't answer them explicitly.

**Comment by hdgarrood** [comment]
I'm confused - I don't understand how to usefully respond without answering those questions. I also think they're good questions which deserve responses if we are to make progress on this front.

**Comment by keijokapp** [comment]
I'm new to PureScript. I quit writing Haskell - both professionally and for personal projects - primarily because of its terrible module system. And that's exactly the same reason I haven't yet got started with PureScript. Having simple and unopinionated module system, which actually solves problems like name conflicts gracefully, just weights up all FP and powerful type system goodies.

Why not ditch Haskell-style modules entirely and embrace ES6/CommonJS module concepts? After all, PureScript compiles down to CommonJS anyway, so why use archaic and problematic module system on top of perfectly good and proven one?

Basically, throw away all the complexity of "package" and "module name" from language and refer to modules with arbitrary context-specific strings (i.e. filenames). That would make concept of module more abstract, removes namespacing problems, simplifies module system (a lot - well, from developer's perspective anyway), and enables us to build less opinionated tooling, which works on abstract files rather than modules of specific language. I'd imagine it would also make FFI simpler.

Obviously, it's quite a radical proposal with its own problems and I'm really interested in others' opinions.

**Comment by garyb** [comment]
If you'd throw out Haskell for JavaScript just over namespacing then you must have some seriously strong opinions about it. 😉 That seems like a very poor tradeoff to me.

Are you proposing hierarchical dependencies along with that then? Because that's not something that works, but without it I don't know how you'd resolve dependencies of dependencies, if everything has to be a file path.

Even in node what you're describing isn't really how it works, since you depend on things in `node_modules` without relying on path, so it still name collisions and such.

I don't see how what you're suggesting would have any effect on the FFI.

**Comment by keijokapp** [comment]
I love Haskell as a language but there was a bit more than just namespacing.

Sorry for not being clear. I really meant to ask just one question: if PS is moving away from global module namespace by introducing a syntax extension, which makes module names package-local, then do we really need excplicit module names at all?

I'm not proposing supporting nested dependencies. Not *every* module reference needs to be a trivially resolvable file path - it might be e.g. `prelude/maybe.purs` and resolver maps it to suitable source file. Package name conflicts can still occur.

It was a wild thought that since modules are referenced by strings, we could just `import` any CommonJS compatible file directly and provide type information in place. That's roughly what I meant by more abstract modules and more generic tooling.

**Comment by srghma** [comment]
related discussion https://discourse.purescript.org/t/the-only-thing-i-hate-in-purescript-and-haskell-imports-discussion/1700

**Comment by wclr** [comment]
I'll place the link on the related topic too: https://discourse.purescript.org/t/proposal-on-purescript-package-management/2245

**Comment by rhendric** [comment]
Seeing as I sorta just scuttled an otherwise promising PR on the premise that it should be blocked on this discussion resolving, I feel it behooves me to try to work on that somehow.

Let me see if I can summarize the state of play:

### Terminology

I will try to use these terms when describing the problems, to add specificity to where the concepts conflict and to highlight that changing the meaning of ‘module’ or ‘package’ doesn't itself solve any of these underlying issues.

* **namespace**: a unique identifier which becomes part of the global name of a declaration, so that references to the declaration can use a short name which resolves to the correct full name
* **compilation unit**: the smallest unit of code that the compiler must reprocess if a change occurs within it
* **publication unit**: the smallest unit of code that can be shared, versioned, and depended on by downstream code
* **encapsulation unit**: the largest unit of code that can contain a declaration that is not globally accessible
* **workspace**: the code under active development by the user, as distinct from upstream dependency code
* **project**: a collection of code expected by a user to compile correctly and interoperate, generally consisting of a workspace plus dependencies
* **package set**: a collection of specific versions of interdependent publication units which have been verified to collectively compile successfully

### PureScript 0.14

In PureScript today, a **module** is a single file that acts as a **compilation unit** and an **encapsulation unit**. A module includes a name which acts as the **namespace** for all of the code in the module. No relationship is imposed between the name or location of the file and the name of the module. Modules cannot depend on each other cyclically.

Spago and older PureScript package managers consider a **package** to be a collection of **modules** with its own name. Package names are not **namespaces** and do not interact with namespaces. Packages do act as a **publication unit**. Package managers give packages a way to specify which other packages that they depend on, at which versions, and from where to obtain those packages.

The PureScript compiler processes all the code in a **project** homogeneously, from source. It doesn't differentiate between **workspace** code and dependency code installed by package managers.

Spago uses a **package set** to resolve **package** versions and prevent conflicts between different dependencies.

### List of complaints

#### Library authoring problems

101. Module names need to be unique in a project, which practically means globally unique. Our existing culture and tooling doesn't facilitate this. #2437
102. There is no relationship between the namespace used for a referenced declaration (module name) and the name needed to make that declaration available in a project (package name). #2437
     1. This means, for example, that going from reading code to including that code in one's own project involves manually searching for the right packages to install. #2437
     2. This also means that it is easy to reference code in a publication unit that is not included in what the package manager thinks are the dependencies of a project, if the referenced code is depended upon transitively. This in turn causes the full transitive closure of a project's dependencies to be de facto part of that project's public API, for purposes of versioning. [#3493 (comment)](https://github.com/purescript/purescript/issues/3493#issuecomment-449386206)
103. The fact that encapsulation units can't span multiple namespaces means that library authors sometimes need to create `*.Internal` namespaces which by convention should be ignored by consumers. [#3493 (comment)](https://github.com/purescript/purescript/issues/3493#issuecomment-449275687)
104. The fact that namespaces can't depend on each other cyclically means that sometimes code needs to be divided into more namespaces than would otherwise be natural (viz. `*.Types` modules). [#3493 (comment)](https://github.com/purescript/purescript/issues/3493#issuecomment-449289859)
105. There is a convention that links namespaces and file paths, which is optional.
      1. But if followed, it means the module name specified in the file is redundant. [#3614 (comment)](https://github.com/purescript/purescript/issues/3614#issuecomment-487756427)
      2. There is no way to enforce that it is followed. #3614, but see [#3614 (comment)](https://github.com/purescript/purescript/issues/3614#issuecomment-487793939)
106. Namespaces appear to be hierarchical, but aren't.
      1. Relative imports aren't a thing. Writing the entire namespace in every file is redundant, with all the drawbacks redundancy implies. [Discourse](https://discourse.purescript.org/t/the-only-thing-i-hate-in-purescript-and-haskell-imports-discussion/1700)
      2. There is no relationship between the hierarchy implied by the namespace and publication units. [Discourse](https://discourse.purescript.org/t/proposal-on-purescript-package-management/2245)

#### Missing optimizations

201. A larger compilation unit would enable more aggressive optimizations that can't be done by looking at one file at a time. [#3493 (comment)](https://github.com/purescript/purescript/issues/3493#issuecomment-449275687), but :-1: from @kritzcreek in [#3493 (comment)](https://github.com/purescript/purescript/issues/3493#issuecomment-449322204)
202. Workspace code and dependency code change at very different rates, but the compiler can't apply different rebuilding strategies to each. [#3493 (comment)](https://github.com/purescript/purescript/issues/3493#issuecomment-567337683)
203. Publication units can only be consumed in their uncompiled form; there is no way to depend on a built publication unit. #2477

#### Desired features

301. It is not possible to use multiple versions of a publication unit in a project, because each version reuses the same namespaces. [#3493 (comment)](https://github.com/purescript/purescript/issues/3493#issuecomment-451535514), but a soft :-1: from @hdgarrood in [#3493 (comment)](https://github.com/purescript/purescript/issues/3493#issuecomment-506469441)
302. There is no place in the compilation pipeline to inject any build steps (code generation, for example) specific to a publication unit, to be used by all downstream consumers of that publication unit. [#3493 (comment)](https://github.com/purescript/purescript/issues/3493#issuecomment-451535514), but :-1: from @hdgarrood in [#3493 (comment)](https://github.com/purescript/purescript/issues/3493#issuecomment-453773750)
303. We want to suppress warnings the way that `purescript-psa` does, by only looking at warnings generated in workspace code. This requires the compiler to know what the workspace is. #4156
304. There's no way to depend on a collection of related publication units by referring to them all with one name. Granular publication units always means lots of entries to manage in a project's dependency list. [Discourse](https://discourse.purescript.org/t/proposal-on-purescript-package-management/2245)

#### Miscellaneous objections

401. Philosophically, needing two distinct things called packages and modules in the ecosystem is a smell, and maybe we should have a better handle on why that is and if it should be. [#3493 (comment)](https://github.com/purescript/purescript/issues/3493#issuecomment-602309510)
402. The mismatch between PureScript's namespaces and the namespaces used by the backend language makes PureScript less of a first-class member of the target language ecosystems. [#3493 (comment)](https://github.com/purescript/purescript/issues/3493#issuecomment-621065434)

Please point out any substantial complaints that I didn't include and I'll add them.

### Notable proposals

1. Move to the convention that namespaces in a single publication unit share a common, unique prefix, with an obvious mapping to the publication unit name. (`Effect.*`)
    * Addresses 101, 102.i, and 106.ii.
    * Suggests an approach to 303 of filtering by module name prefixes.
    * Requires zero implementation work in the compiler.

2. Extend namespaces from unstructured strings to a pair of strings (package name, unique name within package). Reflect this in import syntax by requiring (optionally?) imports to include package names for imports outside of the current workspace. Module declarations do not need to include the package name; the compiler associates directories with either the workspace or an individual dependency.
    * Addresses 101, 102, 106.i (partially), 106.ii, 202, and 303.
    * Requires a syntax change to the language.
    * Requires directory structure to be configured somehow.

3. Allow cyclical module dependencies within a publication unit by compiling cycles as one unit. The compiler associates directories with a publication unit.
    * Narrowly addresses 104.
    * Requires some clever implementation work to make sure that the right things are recompiled.
    * Requires directory structure to be configured somehow.

4. Just tell the compiler how to distinguish between workspace code and dependency code.
    * Narrowly addresses 202 and 303.
    * Pretty straightforward to implement if we don't want to do anything else on this list.

5. Couple the encapsulation unit to the publication unit instead of the namespace. Implement a syntax for publication-unit-private declarations. The compiler associates directories with a publication unit.
    * Narrowly addresses 103.
    * Requires a syntax change to the language, obviously, but likely not a breaking one.
    * Requires directory structure to be configured somehow.

6. Implement relative import syntax.
    * Narrowly addresses 106.i.
    * Requires a syntax change to the language, obviously, but likely not a breaking one.
    * Probably very straightforward to implement if done on its own.
    * Possibly not necessary if we take a different approach.

7. Collapse the namespace, encapsulation unit, compilation unit, and publication unit down to the same entity. Modules are packages, and as such are published and versioned individually. Dependencies are read from the imports of a module and downloaded from a registry by that name. Solve the ensuing version hell with... package sets, I guess? Or some sort of configuration?
    * Addresses 102, 304, and 401.
    * Weirdly addresses 101 by solving it directly in the registry.
    * Having single-module packages is probably fundamentally incompatible with other proposals that depend on a publication unit being larger than a namespace.
    * Is kind of bonkers.

---

I'm sure there are more proposals worth writing up in this format, but I have to call it a night for now. I hope to come back to this. I also hope this is helpful.

**Comment by f-f** [comment]
Thank you for writing this down @rhendric! Having a big recap in this format is indeed helpful.

I'd just like to add that I see this feature to be very tied together with the overall direction/design of [the Registry](https://github.com/purescript/registry), since they are both about publishing/compilation units.
It's important to note that while the overall design is fairly stable at this point and the implementation of the publishing pipeline is there, the whole thing is not deployed yet (but we'd all like to do that asap).
This means that __now__ is probably the best moment to stabilize this feature and inform both designs, since we're still able to easily change Manifest schemas and other implementation details, while it's going to be much harder later.

**Comment by JordanMartinez** [comment]
Just wanted to say that there's a lot here, and I'll need to read through it a couple of times. So, inactivity here is not because it's been forgotten.

**Comment by JordanMartinez** [comment]
My response has two parts:
1. Above, @rhendric described the problems, which tend to focus on the negative aspects of things. So, I want to invert that, so that this issue describes the outcomes that a solution would achieve as I think that will help us identify next actions.
1. Respond to the notable proposals

<details>
<summary>1 - Outcomes solution must fulfill</summary>

- 101/102 - Make declarations include the package's name in their namespace. Seems like the `"package".Module` syntax would be the approach with the least amount of breakage
- 103/104/201 - make encapsulation units "larger" in that sub units can talk to one another without being visible/accessible by external units. In other words,
    - either enable single files to have multiple modules where modules in that file can have interdependencies (works, but likely has other issues)
    - or specify that a given directory is an encapsulation unit where the single-module files in that directory can have interdependencies (works, but then requires a separate file to indicate which declarations within those files are 'exported' by that "directory module").
- 105/106/201 - require encapsulation units names to match their file path
    - this could be done whether aggressive optimizations were done or not.
    - Windows might have issues with this due to having constraints on how long file paths can be.
    - I can adjust my learning repo to use some other convention (e.g. `01-02` becomes `ab-ac`) and still achieve my intended effect (e.g. auto-sorting files in the order in which I want them read and producing a ToC easily).
- 202/203 - Make "precompiled dependencies" possible, so that the compiler does not need to recompile them from source
    - Do "precompiled dependencies" imply the final backend-specific output (e.g. JavaScript), the intermediate language output (i.e. `CoreFn`), or both options?
        - If the first, then the compiler's work would speed up significantly, but would that affect usages of `runSupplyT` where we thread a counter variable through various computations?
        - If the second, is this a good temporary solution until the first can be implemented?
        - What about situations where the `src` directory is the main code to compile in one situation and a dependency in another situation (e.g. tests, benchmarks, etc.)?
- 303 - indicate which globs are "precompiled dependencies," which are "local dependencies," and which are the "main source code."
    - This could be done using two CLI args: `--precompiled DIR1,DIR2` and `--local-libs DIR1,DIR2`:
        - Given a `.spago` folder of precompiled dependencies and no "local dependencies", one would compile the `src` folder.
            - `purs compile --precompiled ".spago" ".spago/*/src/**/*.purs" "src/**/*.purs"`
        - Given a `.spago` folder of precompiled dependencies AND a `src` folder of "local dependencies," one would compile the `test/` (or `benchmarks/` or `examples/`) folder.
            - `purs compile --precompiled ".spago" --local-libs "src" ".spago/*/src/**/*.purs" "src/**/*.purs" "test/**/*.purs"`
    - Depending on how this feature interacts with 202/203, we might need to also indicate where the FFI files are. For example, if a "precompiled dependency" was the `CoreFn` output, perhaps the globs passed into the compiler would also need to indicate what the corresponding folder is for that globs' FFI files.
- 304 - this sounds more like a registry issue
    - For example, on top of publishing package `foo` and `bar`, one could potentially publish a 'meta-package' `foobar`. A 'meta-package' would function like a mini package set: there are a group of related versioned packages and installing a meta-package would be the same as installing all of its packages at their versions. Potentially, one could override the metapackage's version of `foo` to something else within the project's dependencies (similar to overriding a version for a given package in a Spago project) or drop ones that don't matter. This could be how we make maintenance easier on `core` libs by storing them in a monorepo, and still make each package (as it is now) granular, and still enable a `spago install base` command that gets all core libs in one easy command
- 301/302 - I'm not sure where I stand on this, so I'm leaving it out.
- 402 - I don't understand this

Responding to 401

> Why don't we want multiple PureScript files to work together as a single unit that you cannot "see through?

I would like this. If I'm understanding this idea correctly, one way to express this idea would be to have something "larger" than a module but "smaller" than a package. This yet-to-be-named entity would
- enable modules to have interdependencies and talk to one another
- have their own declaration export list, so that types defined in modules cannot be imported outside of this entity (effectively removing the need for `*.Internal` modules).

Correct me if I'm wrong, but isn't sort of what Rust does?

</details>

<details>
<summary>2 - Proposals</summary>

In short, I think we should focus on Proposal 3 and 5.

1. If module names in a file (e.g. `module Foo where`) are redundant, then this doesn't seem like the best approach.
2. I think the import syntax would be required to include the package name, so that in cases where the `src` directory is the source directory in one case (building the library code) and is itself a "local-lib" dependency in another case (e.g. running tests) still works in both situations.
3. This seems like it would address all the points above. What would be the thorny design issues to work out?
4. This seems like a good temporary solution until something more fleshed out is designed and implemented. However, "temporary" means the registry would always have this "design baggage" since the registry will always be backwards-compatible.
5. Could you clarify this further? Perhaps by providing an example?
6. While this can solve one part of our problems, I don't think it solves all of them and is thus not the best solution.
7. The original thread that raised this idea also indicated that it wasn't meant to be taken seriously.

</details>

**Comment by hdgarrood** [comment]
In my mind allowing cyclical module dependencies has a pretty severe drawback which I haven’t seen mentioned, in that backends which don’t allow cyclical module dependencies would have to do some very awkward gymnastics to figure that out. But I also think we should consider cyclical dependencies out of scope for this issue; I don’t see a need to address this together with package awareness, and also there’s already a lot here and I think that #if we are not very careful about keeping the scope narrow we will never get anywhere.

**Comment by garyb** [comment]
Just a quick note on cyclic module dependencies also: I guess it's only a problem if we're saying that PureScript module artefacts are supposed to be 1:1 with artefacts in the relevant backend, since the easy solution is just collapsing the modules into one.

**Comment by hdgarrood** [comment]
Speaking of scope reduction, I think we should also ignore 304 (collections of packages) for now. I think it’s extremely likely that implementing this would add unnecessary complexity and create more problems than it solved, but in any case it’s something we ought to be able to add later after making some progress on the question of compiler package-awareness.

**Comment by hdgarrood** [comment]
I’m uneasy about the idea of proposal 5 because of the difficulty of configuring which packages live where and also of keeping that configuration consistent across different tools or compiler invocations. I would lean more towards creating a concept of a “package database” which could include lots of packages at different versions, as well as precompiled versions of those packages. That way, I think we can help avoid footguns by keeping the format of a package database internal to the compiler, and saying that you can only interact with a package database via a CLI that the compiler provides.

**Comment by JordanMartinez** [comment]
> In my mind allowing cyclical module dependencies has a pretty severe drawback which I haven’t seen mentioned, in that backends which don’t allow cyclical module dependencies would have to do some very awkward gymnastics to figure that out.

Just curious. Are there any current backends that are examples of this?

So, if my "solutions" for 103/104/201 and 304 were completely dropped, would the remaining solutions proposed be an appropriate enough scope for this effort?

> That way, I think we can help avoid footguns by keeping the format of a package database internal to the compiler, and saying that you can only interact with a package database via a CLI that the compiler provides.

Sounds good to me.

**Comment by rhendric** [comment]
I don't have time to write a long thing right now, but I've been mulling over the problem and I'm starting to converge on an opinion, the short version of which is to start with proposal 2, which I assert has nice expansion paths to cover the problems it doesn't directly address. It'll probably be a few days before I can come back to writing this all out though.

**Comment by rhendric** [comment]
### Proposal 2A

As the name would suggest, this is an elaboration of proposal 2. I had been writing a whole long justification about why this is where I've settled for now, but rather than spend too much more time on that I think I'd rather put this out there just as a spec and see what people who aren't the voices in my head think.

This proposal:
* introduces the concept of a ‘package’;
* introduces the concept of a package mapping file;
* specifies new behavior for `purs compile` when provided with a package mapping file;
* extends the syntax for import statements;
* introduces `Default` as a module name of interest for packages, much like `Main` is a module name of interest for programs;
* recommends an ecosystem-wide restructuring of module naming conventions;
* and finally suggests a transition plan and some additional features to assist with this restructuring.

Proposals 1 and 4 are not necessary if this proposal is implemented. Proposals 3, 5, and 6 remain as possible future work that would be compatible with (and likely facilitated by) this proposal. This proposal is not compatible with proposal 7; I think there's no reason to force publication units and namespaces to coincide, and I can spill more words on this opinion on request.

#### Packages

Under this proposal, a package is a new abstraction consisting of:
* a name, unique across the ecosystem (by convention, made of lowercase Latin letters and hyphens)
* a path from which `purs` can load modules (this could in principle be things like zip files or configuration files which point to other locations on disk, but for the rest of this proposal, I will consider these paths to be directories)

Not every module belongs to a package, but every module loaded via the new package resolution procedure (defined below) is considered to belong to the package being resolved.

The name of a package is used to qualify all of the modules it contains. In other words, the namespace for a top-level identifier goes from having type `ModuleName` to having type `(Maybe PackageName, ModuleName)`.

Note in particular that this means that modules defined in different packages are now allowed to have the same module name, because they will have different package names, and thus will define different namespaces.

#### Package mapping file

A package mapping file is a simple JSON file that maps package names to package paths. This proposal doesn't require a particular name for this file, but as a convention, I recommend `purs-packages.json`.

An example package mapping file might contain:
```json
{
  "console": ".spago/console/v5.0.0/src",
  "effect": ".spago/effect/v3.0.0/src",
  "prelude": ".spago/prelude/v5.0.0/src"
}
```

Paths in the mapping file, if relative, are relative to the directory containing the file.

Importantly, this mapping file is intended to contain all of the packages used by the compilation, not just the direct dependencies of a project.

I expect that package managers like Spago will write to this file, but it should also be possible to manually create or edit such a file and use `purs` without a package manager.

#### `purs compile --packages`

The `purs compile` command will have a new command-line option called `--packages`, which takes the path of a package mapping file as an argument. When this option is provided, `purs compile` will run this package resolution procedure:

* For each (name, path) in the package mapping file:
  * Resolve the path to an absolute path using the location of the package mapping file.
  * If the absolute path is a file, issue an error (this is reserved for possible future extensions).
  * If the absolute path is a directory, recursively find all `*.purs` files under that directory and add them to the list of files to be compiled.
  * Any modules loaded from files thus added should be considered qualified with this package name.

Files provided to the `purs compile` command directly are still loaded as usual; the modules contained therein are not considered qualified with any package name.

#### New import statement syntax

The existing syntax for import statements is:

```
ImportDecl := 'import' ModuleName Imports ( 'as' ModuleName )?
Imports := ε | '(' Import (',' Import)* ')' | 'hiding' '(' Import (',' Import)* ')'
```

This proposal extends this syntax to:

```
ImportDecl := 'import' ( ModuleName | PackageName ModuleName? ) Imports ( 'as' ModuleName )?
PackageName := StringLiteral
Imports := ε | '(' Import (',' Import)* ')' | 'hiding' '(' Import (',' Import)* ')'
```

The semantics of the various forms of `ImportDecl` break down as follows:
* `'import' (m :: ModuleName) Imports ...` imports from the namespace `(p, m)`, where `p` is a `Maybe PackageName` representing the package name of the current module, if any
* `'import' (p :: PackageName) (m :: ModuleName) Imports ...` imports from the namespace `(Just p, m)`
* `'import' (p :: PackageName) Imports ...` is sugar for `'import' p 'Default' Imports ...`, which brings us to the next section...

#### The Default module

The above desugaring is the only thing that's special about the name `Default`. It's not a reserved word or anything like that. It's just a module name like any other, except that it can be omitted from import statements when importing across packages.

#### Module naming conventions

This proposal recommends that new libraries follow the following module naming conventions:

* Do give your library a `Default` module, and export or reexport your basic (or your entire) API from that module.
* Do use an `Internal.` prefix on modules which are not meant to be a part of your API.
* Don't use a `YourProjectName.` prefix on your modules. The package name will serve this purpose.
* Don't use `Control.`, `Data.`, or other Haskellish module prefixes for your modules.

Putting it all together, this is the transformation I ultimately expect PureScript code to make:

```purs
module Main where

import Prelude

import Effect (Effect)
import Effect.Console as Console
import Data.HashMap as M
```

to

```purs
module Main where

import "prelude"

import "effect" (Effect)
import "effect-console" as Console
import "unordered-collections" HashMap as M

-- Remember that this is equivalent to
-- import "prelude" Default
--
-- import "effect" Default (Effect)
-- import "effect-console" Default as Console
-- import "unordered-collections" HashMap as M
```

#### The transition plan

This is a big change. From the above, it's pretty clear that this will eventually involve rewriting every PureScript file that exists. The good news is that we can do this in stages, and the rewrites are mostly mechanical.

First, because this proposal is strictly additive, on release day, nothing breaks. If Spago and other callers of `purs` continue to provide all source files as arguments to `purs compile`, they will all be considered ‘unpackaged’, and nothing about how they're defined or imported will change. Beyond release day, I envision that Spago will have some way of identifying a particular package version as unreformed or reformed (marker file in the source? property in the package set data?); reformed packages will go into a package mapping file, and unreformed packages will continue to be provided as arguments to `purs compile`. As long as everything upstream of the reformed packages is also reformed, a package set can contain a mix of reformed and unreformed packages and still be usable, and end users only need to rewrite the imports for the packages that have converted. So this can be rolled out in waves across the ecosystem if we want.

Better still, while a library author may need to think a little bit about how to restructure their module names in the new paradigm, consumers just need to mechanically replace `import Old.Module.Name` with `import "package-name" New.Module`. I propose developing a tool to do this replacement automatically. We could either have a central registry that contains a list of `"Old.Module.Name" -> ("package-name", "New.Module")` mappings, or use a temporary `-- formerly Old.Module.Name` pseudo-pragma that the tool recognizes at the top of module files, and the tool could scan all the paths used in a package mapping file for such pragmas and build the mapping from that (pulling the package name from the package mapping file, the old module name from the pragma, and the new module name from the `module` declaration). Either way, fixing ModuleNotFound errors associated with this paradigm shift would be as easy as running the tool, which could be done repeatedly and wouldn't require any user intervention.

**Comment by JordanMartinez** [comment]
I like this. It seems to address the more painful parts of modules without going so far that we can't later add other proposals that aren't yet fleshed out. I really like the transition plan.

It also sounds like it would address another issue commonly encountered with Spago. If I want to run my tests, I either add the test dependencies to `spago.dhall` or add them in a separate `test.dhall` and call `spago -x test.dhall test`. With this approach, each library or application could have its own package (e.g. `packageName`) and its test/benchmark/examples modules could have their own package (e.g. `packageName-tests`, `packageName-benchmarks`, `packageName-examples`) that depend on the main library/application's package.

At the same time, tools like Spago would need to be careful about managing the `purs-packages.json` file in the above situation. If I only build the `src` directory and its dependencies, the tool might think it should remove other entries in that file (e.g. the `packageName-tests` package).

Also, how would the `purs-packages.json` file work when one is using both Spago and Pulp? For example, managing their dependencies via Spago but publishing via Pulp? When Spago is managing the file, package `foo` should point to `.spago/foo/v3.0.0/src`. When Pulp is managing the file, package `foo` should point to `.bower_components/foo/src` (or whatever the path would be for Bower). Would we tell `purs compile` via a CLI arg where to find this `purs-package.json` file?

Lastly, since we already have a `package.json` file for NPM projects, and a `packages.dhall` file for Spago, could the name convention for `purs-package.json` be something that doesn't quite have `package` in the file name?

**Comment by rhendric** [comment]
I imagine the package mapping file as a low-level interface; if you're using Spago, then Spago generates it and controls its contents, and gives it whatever name Spago thinks best. In that scenario, I wouldn't add it to source control, for example. (In fact, maybe Spago puts it in the `.spago` directory.) Similarly for Pulp. (And yes, the proposal says the file should be specified by the `--packages` CLI arg; there is no default place to look for one.)

If we recommend that people create multiple local packages for the main library, tests, benchmarks, and examples, then we have the problem again that PureScript can't automatically disambiguate between workspace code and library code, which we want to do for compilation optimizations and warning suppression. In this proposal, I want being qualified with a package name to effectively mean ‘came from somewhere else’, and anything not qualified is the workspace. I'm not sure what's so bad about the `test.dhall` pattern, though; I thought it does the trick well enough, although I see it's a little unintuitive to people expecting a more `devDependencies`-like interface. It doesn't seem like it saves any configuration to set up a bunch of smaller projects; they'd all need their own Spago configurations either way, right?

**Comment by JordanMartinez** [comment]
> In this proposal, I want being qualified with a package name to effectively mean ‘came from somewhere else’, and anything not qualified is the workspace. I'm not sure what's so bad about the `test.dhall` pattern, though

That makes sense and it does make things simpler. The `test.dhall` pattern isn't bad, just annoying.

**Comment by JordanMartinez** [comment]
Another question going in a slightly different manner. You are proposing we use a `purs-package.json` file to map packages to source globs. The library-consumer would have to maintain that file, whether manually or via a tool like Spago. But, is there a way for this to be figured out automatically? If there was, what tradeoffs would that entail?

For one possible automatic approach, why wouldn't this be specified within the package itself? 

For example, why wouldn't `purescript-prelude` have a file within its repo that indicates that all of its modules belong to the `prelude` package? Rather than forcing library-consumers to indicate what the package-module mapping is, why not force library-producers to do that? This could be a special file called `purs.package` (or something) that looks like this:
```
package "prelude" where

-- if we wanted to specify any package-specific thing
-- in the future, we could put it in this file but this file
-- would be blank besides the package declaration above
```
If I pass in the source glob `.spago/prelude/v5.0.1/src/**/*.purs`, the compiler could automatically check to see if there is a `.spago/prelude/v5.0.1/src/purs.package` file and use that to figure out that to which package all of the modules within that source glob belong. 

In case that's too flaky, perhaps we could do the following:
1. Run `purs compile ".spago/prelude/v5.0.1"` (the compiler will assume that this is a directory and that a `purs.package` file will be within this directory (e.g. `.spago/prelude/v5.0.1/purs.package`)
1. The `purs.package` file tells the compiler the source globs for this package

```
-- | package documentation
package "prelude" where

-- In other words, all modules found in the below globs
-- should be considered part of this package.
-- Glob paths are resolved based on
-- the location of this file.
source-modules "./src/**/*.purs"
```

1. To disambiguate between "precompiled" dependencies and "workspace" code, perhaps the `purs.package` file just states that directly

```
package "prelude" where

-- source modules if one wants to compile from source
source-modules "./src/**/*.purs"

-- compiled modules' externs that can be used
-- by the `purs compile` command
compiled "./output/**/*.js"
```

**Comment by JordanMartinez** [comment]
Another question inspired by my the above package declaration's `compiled` field.

Regardless of whether the `purs.package` file idea is accepted or not, would it be helpful to distinguish one package from another in the `output` directory? Instead of `output/Module.Path/` as the directory that contains the output of the module, `Module.Path`, would it now be `output/packageName/Module.Path/`? If so, how would that work in situations where modules are not qualified with a package name?

**Comment by hdgarrood** [comment]
I like this proposal, although I have one suggestion about the specifics of the interface, because I think a drawback here is that it is not entirely clear (at least, not to me) how it could be extended to solve the issue of precompiled dependencies safely.

As an alternative to the package mapping file, I'd suggest:

- A package database is a directory on the filesystem somewhere. Its structure is considered an implementation detail; nothing other than `purs` should be poking around inside one. The compiler chooses a sensible default location for the package database and uses that package database for most commands, unless requested otherwise; perhaps a `--package-db DIRECTORY` option.
- Within a package database, there are a number of packages, each of which is identified by a package identifier, which is a pair of a package name and a version. (Note that this suggests that only published packages in the central registry ought to be added to a package database.)
- You can add a package to a package database by running something like `purs pkg add DIRECTORY`, where the specified directory should contain a package manifest file which allows the compiler to figure out the name, version, and dependencies of the package. This command is considered low-level and intended to only be used by tooling such as Spago.
- When tooling (such as Spago) installs dependencies for a project, it would also assemble a JSON file which contains a list of package identifiers for the packages that this project depends on, which might look something like `["prelude-5.0.1", "console-4.0.0"]`. That file would be passed to each invocation of commands like `purs compile` or `purs repl`. Given this information as well as the location of the package database, the compiler should be able to identify where the appropriate sources are.
- Within a package database, in addition to package sources, there is also a nix-style store for precompiled dependencies. By "nix-style", I mean that all build products are stored in a directory which includes a hash of the library itself as well as the dependencies it was built against. Note that we might need to store different precompiled versions of a given package at a given version, because if package `a-1.0.0` depends on package `b` and can be built against both `b-1.0.0` and `b-1.1.0`, then `b` might have changed between those two versions in a way that is reflected in the build products of `a` - maybe an instance name has changed - and so we have to make sure we select the right one according to the version of `b` we are currently using.

> Regardless of whether the purs.package file idea is accepted or not, would it be helpful to distinguish one package from another in the output directory? Instead of output/Module.Path/ as the directory that contains the output of the module, Module.Path, would it now be output/packageName/Module.Path/? If so, how would that work in situations where modules are not qualified with a package name?

I think it would be necessary to include the package name in the output directory, because in this context we are no longer requiring module names to be globally unique - for example, most packages would be expected to define a module `Default`. We might want to use a scheme like `output/packageName__Module.Path` though, so that existing setups which expect to find JS files matching globs like `output/*/*.js` don't break.

**Comment by garyb** [comment]
Bikeshedding comment: why a string literal for the package name? It seems like just using the package name would be fine too, it's lexically unambiguous with a non-package import if the convention rules you suggested are instead made a requirement. Making them a requirement seems preferable since it ensures that the names will be compatible across filesystems in a way that might not be true if package names were just allowed to be anything that the local OS supports as a directory name.

**Comment by rhendric** [comment]
@JordanMartinez 

> Rather than forcing library-consumers to indicate what the package-module mapping is, why not force library-producers to do that?

Could work—it's worth noting that library producers already specify a package name in `bower.json`, and presumably in some configuration file when we no longer publish to Bower. But I think that possibly demonstrates that this is mixing concerns in a bad way. On the library producer's end, the package name is just a publication concern; `purs` doesn't need to know it. So it naturally lives in the configuration of whatever tool does the publishing. On the consumer's end, the package name is something `purs` does need to know, and maybe the consumer has a reason to use a name other than the name it was published under. (Perhaps a benchmarking script wants to compare the performance of multiple versions of a package, say.) I think it's more natural and more flexible to add this information to the interface between the consumer and their invocation of `purs`, rather than an interface between library author and the consumer's invocation of `purs`.

@hdgarrood

> it is not entirely clear (at least, not to me) how it could be extended to solve the issue of precompiled dependencies safely.

I think I might not have a good enough handle on what the pitfalls are around precompiled dependencies to understand why a package database would be necessary. I had figured that a package manager could drop precompiled dependencies alongside uncompiled dependencies, and `purs` could just copy them over into `output` (or do whatever else). I could see two approaches to identifying dependencies as precompiled: either the presence of certain files in the folder; or changing the package path from, e.g., `".spago/console/v5.0.0/src"` to `".spago/console/v5.0.0.zip"` or `".spago/console/v5.0.0/precompiled.json"`—this is what I meant in the proposal when I said that the case of a package path being a file is reserved for future extensions. Having a global dependency store might be a fine thing, but I would want Spago or whatever package manager to implement and manage that, and then the package mapping files would look like `{ "console": "/path/to/store/9cdahaxba16ba98ighxtf160fl2dhxao-console" }` or whatever—I don't see why that functionality needs to live in `purs`.

> I think it would be necessary to include the package name in the output directory, because in this context we are no longer requiring module names to be globally unique - for example, most packages would be expected to define a module `Default`. We might want to use a scheme like `output/packageName__Module.Path` though, so that existing setups which expect to find JS files matching globs like `output/*/*.js` don't break.

Yes, exactly this! (We only need one underscore as a separator, or really any character that isn't valid as part of a module name, to ensure uniqueness; but maybe two is more readable and this is pretty bikesheddy anyway.)

@garyb

> Bikeshedding comment: why a string literal for the package name?

A few reasons. First: something like `import prelude` really makes `prelude` look like it's an identifier like the other lowercase words that aren't keywords that occur in a PureScript file. But of course it's something fundamentally different. I think the syntax ought to reflect that. Second: this gives us, I think, more flexibility to add things like versioning or namespaces to package names in the future. Sure, we could always extend the syntax in `purs` to support more characters, but I want external tools also to be able to parse package names from import statements (both for the transition strategy I outlined in the proposal, and also for ideas like detecting when the code uses packages that aren't specified as direct dependencies—ideally, the package manager would do this, because the package manager owns the concept of a ‘direct dependency’ in the first place). `'import' StringLiteral` is an easier thing to expect those tools to parse now and in the future than `'import' KebabCasedIdentifier`, oops now it's `'import' ( Namespace '/' )? KebabCasedIdentifier`, oops now it's `'import' ( Namespace '/' )? KebabCasedIdentifier ( '@' Version )?`. Third: consistency with GHC's syntax doesn't hurt—this shouldn't be the most important factor, but it's a good tiebreaker. Fourth: package names don't have to be filesystem-compatible or filesystem-unique; only package paths do. Package managers should decide what the package path will be; they can escape the names, they can use hashes, they can do whatever. `purs` just has to take the string in a source file and look it up in the mapping file.

**Comment by wclr** [comment]
First, thanks @rhendric that you did find the time to analyze the situation and propose a _concrete_ possible solution, this definitely should move things forward. Though I find that there are some complications involved in this proposal that could generally be avoided. I don't know if while your analysis you did familiarize yourself with the ideas I tried to express in [my proposal on the same problem on the discourse](https://discourse.purescript.org/t/proposal-on-purescript-package-management/2245/) a while ago.

I probably should start with a simple and evident thing. I wonder if I am the only person here who _feels_ that such change:

**from:**

```purs
module Main where

import Prelude

import Effect (Effect)
import Effect.Console as Console
import Data.HashMap as M
```

**to:**

```purs
module Main where

import "prelude"

import "effect" (Effect)
import "effect-console" as Console
import "unordered-collections" HashMap as M
```

would be at least a very unpleasant (if not to say awful) thing by itself? It feels like things are moving from **simplicity, purity, strictness, and structure**  towards something like a JavaScript land.

Form follows content, and just by looking at the form of such change one may guess out in what way the underlying things are going to change too. 

This is far from emotional or pure aesthetics judgment, I tried to elaborate on this topic in detail in [the discourse thread](https://discourse.purescript.org/t/proposal-on-purescript-package-management/2245/14?u=wclr) (though my views on some things might change from that time).

I'm still going to address in detail the points of the above proposal, though for now, I would like to appeal to the participants of the recent discussion (@rhendric, @hdgarrood, @garyb, @f-f, @JordanMartinez) with the following question:  

Wouldn't it be better if we could keep the current (simple and strict) semantics of module imports (and even simplify it, for example by getting rid of ubiquitous and low-value namespaces like `Data`, `Control`), at the same time introducing a solid package system basis and the compiler package awareness? Maybe, for now, it is not about what you _think_ but how you _feel_ about it?

**Comment by hdgarrood** [comment]
The main issue with precompiled dependencies is that is only safe to reuse precompiled dependencies for some package across two different projects if _all_ of the dependencies of that package are at the exact same versions in both projects. So just keeping one set of precompiled dependencies alongside each package wouldn’t work - you need one set of precompiled dependencies per install tree (or at least, per subtree of the install tree with the root being that package).

I would prefer to keep this functionality inside the compiler because I think the compiler ought to be the judge of when it is safe to reuse precompiled dependencies, as I think it will be extremely easy to get this wrong, especially in downstream projects where people are less likely to be familiar with how the compiler output works and how this could go wrong. I would rather not provide a compiler API which is easy to misuse, because while we can probably trust Spago to get it right, that won’t be true of third party tools.

**Comment by hdgarrood** [comment]
The main issue with precompiled dependencies is that is only safe to reuse precompiled dependencies for some package across two different projects if _all_ of the dependencies of that package are at the exact same versions in both projects. So just keeping one set of precompiled dependencies alongside each package wouldn’t work - you need one set of precompiled dependencies per install tree (or at least, per subtree of the install tree with the root being that package).

I would prefer to keep this functionality inside the compiler because I think the compiler ought to be the judge of when it is safe to reuse precompiled dependencies, as I think it will be extremely easy to get this wrong, especially in downstream projects where people are less likely to be familiar with how the compiler output works and how this could go wrong. I would rather not provide a compiler API which is easy to misuse, because while we can probably trust Spago to get it right, that won’t be true of third party tools.

**Comment by garyb** [comment]
@rhendric

> First: something like `import prelude` really makes `prelude` look like it's an identifier like the other lowercase words that aren't keywords that occur in a PureScript file. But of course it's something fundamentally different. I think the syntax ought to reflect that.

Isn't that true of having it as a string literal too? It looks like a value and perhaps even suggest dynamic imports if an identifier was used instead.

Actually... you know what, I was going to make some replies to the other points too, but it doesn't really matter. I raised the question as I knew about `PackageImports` in Haskell and it seemed janky to me there too, but really it's just an aesthetics thing for me. I let the IDE manage my imports for me now and will be continuing to do so in the future so the specific appearance of it isn't a big deal. :wink: Also, Wadler's law.

@wclr

> I probably should start with a simple and evident thing. I wonder if I am the only person here who feels that such change:
> ...
> would be at least a very unpleasant thing by itself? It feels like things are moving from **simplicity, purity, strictness, and structure** towards something like a JavaScript land.

I don't fell that at all. What don't you like about it? Is it the packages being disambiguated from the module namespace? I don't see how that change has anything to do with purity / strictness, and has minimal impact on simplicity.

**Comment by wclr** [comment]
> I don't feel that at all. What don't you like about it? Is it the packages being disambiguated from the module namespace? I don't see how that change has anything to do with purity / strictness, and has minimal impact on simplicity.

There are some aesthetic/wholeness considerations of course, but the problem that I see with such change is deeper. I will need to eventually explain why in more detail. For example, the existing import syntax implies abstraction over the file system and also over a package system used, and moving toward string literal names just reduces the level of abstraction, this eventually could even lead to just dropping explicit module names and just replacing them with file system locations, relative paths 😭 (some people seem to like the idea), and - hello JavaScript.

But first of all, in my view, there is just no need for such a change. We need very carefully consider the value that it brings. It seems to me now that we are trying here generally to replace the root module namespace with a string literal, turning this:

```purs
import Halogen.HTML.Elements (element)
```

to this:

```purs
import "halogen" HTML.Elements (element) 
```

just because we can not figure out how to deal with the existing syntax and existing package-related problems, which in my view are solvable without such changes.

**Comment by hdgarrood** [comment]
> which in my view are solvable without such changes.

We have already considered this carefully and I am quite confident that this is not the case, at least not in a way that I would find acceptable. See discussion in #2437.

**Comment by wclr** [comment]
@hdgarrood 
> See discussion in #2437.

That proceeded from the assumption that *module names should be not unique* in the user's codebase (and actually throughout the ecosystem). But this may be a wrong premise that may lead to a wrong solution. Wrong in those terms that there might be a better eventual situation and solution where module name uniqueness is ensured/enforced/encouraged. 

So maybe it is better to think a bit more carefully in another direction for example how to create such conditions (for module name uniqueness) than how to provide users with the ability to name modules as they wish regardless of anything. I see this as a kind of disrespect and abuse of the module namespaces that stems from insufficient understanding of how to use this feature properly and to the full of its power. 

Though currently, we are in the position where we should use it in the right way (as I see and describe it) and we use it - modules have to be unique now - and there are not many problems with it actually, this really makes things more simple, though there exist some edge cases and **a fear that we will not be able to use it properly in future**. But not using module namespaces this way as the unique language feature is almost equal to throwing it out.

Actually what the latest proposal presented here by @rhendric (as well as other prior suggestions) is trying to make is **to extend the module namespace with the literal string package name** (so we would get eventually things like: `output/packageName__Module.Name`).

What I have proposed a while ago instead of extending module namespaces with a package name, just make some changes to those namespaces (primarily to the understanding of their role) and packages contents, and completely remove the need in a literal string package name as **an independent unique identifier** (though it is still possible to have it in a familiar low-case string literal form for appropriate purposes).

Removing something that is not needed without actual loss of features and functionality should be the desired output for any development process. I'm not sure why you don't like this idea, but you probably have some unresolved/unresolvable concerns about it.

I'm still quite convinced that this could be a better solution than making module names actually not unique or better to say making them unique by prefixing it with a string literal package name. But to be unique across the ecosystem module names don't need this crutch that we currently refer to as "a package name", they have this capability built-in, but the proper use of it seems to require some considerable change in the understanding of module namespaces and packages (though it is not drastic just I would say refined).

This is my personal current perspective on the problem of namespaces, packages, and naming, I assume that I may miss something and interpret things in the wrong way: for example, maybe module names really don't have to be unique on a large scale by proper design and should be allowed to be used as freely as possible?

**Comment by hdgarrood** [comment]
Regarding package names in module names: you’ve made your case but we found it unconvincing, and I (and others) have laid out the case for the planned change as clearly as we can. I think further conversation on this topic probably isn’t going to be fruitful, so I am going to ask you to please drop the subject, at least on this thread, so that we can focus on the other aspects of the design which we haven’t yet figured out.

**Comment by garyb** [comment]
@rhendric do you have any more thoughts on the motivation for having a `Default` module? I could take it or leave it, just wondered if there was something specific you thought of that this would particularly improve.

**Comment by rhendric** [comment]
Especially given that some of the complaints (105.i, 106.i) were specifically about redundancy, I felt that a proposal which asked people to go from, for example,
```purs
import Prelude
```
to
```purs
import "prelude" Prelude
```
would be a hard sell. I briefly considered some alternatives—configuring a per-package module name as the default module somewhere, or using some heuristic to find a module the name of which is the common prefix of other module names in the package—but using a module named `Default` seemed most straightforward, and in line with the precedent set by `Main` modules.

(@hdgarrood, I'm still mulling over https://github.com/purescript/purescript/issues/3493#issuecomment-890417999.)

**Comment by purefunctor** [comment]
I came across this issue last year w/o much knowledge on packaging systems but as I worked with other languages more closely, Rust/Haskell in particular, and looked at stuff like Nix/Guix more in depth, the more that I became confident in the 2A proposal. Basically gathering different perspectives before making a judgment.

A few things I had in mind after another read of this thread:

## string-like syntax

I had a fairly negative opinion of this before, much like I did with `require`/`import` in JavaScript but I'm fairly neutral about it now. It's easier to implement syntactically and is generally more flexible than constraining package names to _only_ be valid identifiers.

Right now, I'm neutral about expanding this import form to allow relative imports, since I feel that this leads into the discussion of local packages and whether or not those should be `purs`-facing or build-tool-facing. 
```hs
import "../my-local-package/"
```

As for an import form that allows for version tags, I feel that this wouldn't be used as much to warrant support for it.
```hs
import "prelude@1.0.0"
```

I think the package manifest `purs-packages.json` should _definitely_ allow aliases though. The need for this would be very niche, for instance with performance snapshot tests but this is _good enough_ flexibility for me personally.
```json
{
  "free": ".spago/free/v7.0.0/src",
  "free-legacy": ".spago/free/v6.0.0/src"
}
```

## module naming conventions

I feel that package-based imports is inherently orthogonal to the naming conventions that we inherited from Haskell. What I mean by this is for stuff like `effect` and `console` for instance. Currently, we have the following form:
```hs
import Effect
import Effect.Class
import Effect.Console
import Effect.Class.Console
```

During the period that we roll out package-awareness, we'd see imports along the lines of the following, which has quite a lot of redundancy:
```hs
import "effect" Effect
import "effect" Effect.Class

import "console" Effect.Console
import "console" Effect.Class.Console
```

And eventually, we would see import declarations of the following form; however, this is not ideal because of the naming convention we inherited from Haskell where we place some module under a path that's already defined from another package. 
```hs
import "effect"
import "effect" Class

import "console"
import "console" Class.Console  -- wait, what do you mean by Class?
```

This is a particularly niche case, and I think there's multiple correct solutions, but what personally makes sense for me is:
```hs
import "effect"
import "effect" Class

import "console"
import "console" MonadEffect
```

Of course, there's also this which is much more neutral, and would really only need modification of the `purs-packages.json` manifest, if I understand correctly.
```hs
import "effect-console"
import "effect-console" Class
```

(:bike::house:) Perhaps instead of `Default` we can use `Lib`/`Library`?

## nix/guix support

I feel that letting Nix/Guix manage fetching packages, and then putting the store paths under the `purs-packages.json` manifest feels intuitive, and that they would _just_ work. In terms of pre-compiled dependencies, how they're built (as in the act of `purs compile` invocation) would be specific to Nix/Guix respectively. Then, it'd be a matter of feeding them into the local project being developed, which runs into https://github.com/purescript/purescript/issues/3493#issuecomment-89041799

Perhaps it's just that it's a new feature, but `purs pkg` being able to mutate a package database somewhere in the file system feels a bit _impure_? I don't know how to put it 😅. I feel that it can live as a separate tool on its own but then that involves building the compiler to yield two executables instead of just one.

I strongly believe that `purs pkg` can integrate quite well with Nix/Guix, much the same way that `ghc-pkg` does.

**Comment by hdgarrood** [comment]
It's true that a Nix-style store would be ideal here, but unfortunately I think Nix should be totally out of the question until it gets first-class Windows support at absolute minimum (WSL doesn't count).
